nemo_text_processing/text_normalization/vi/verbalizers/time.py

```diff
@@ -0,0 +1,149 @@
+# Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pynini
+from pynini.lib import pynutil
+
+from nemo_text_processing.text_normalization.vi.graph_utils import (
+    NEMO_NOT_QUOTE,
+    GraphFst,
+    convert_space,
+    delete_preserve_order,
+    delete_space,
+)
+from nemo_text_processing.text_normalization.vi.utils import get_abs_path
+
+
+class TimeFst(GraphFst):
+    """
+    Finite state transducer for verbalizing Vietnamese time.
+
+    Converts tagged time entities into spoken form, e.g.:
+    - time { hours: "tám" minutes: "ba mươi" } -> tám giờ ba mươi phút
+    - time { hours: "mười bốn" minutes: "mười lăm" } -> mười bốn giờ mười lăm phút
+    - time { hours: "chín" } -> chín giờ
+    - time { minutes: "ba" seconds: "hai mươi" } -> ba phút hai mươi giây
+    - time { hours: "tám" minutes: "hai mươi ba" zone: "g m t" } -> tám giờ hai mươi ba phút GMT
+
+    Args:
+        deterministic: if True will provide a single transduction option,
+            for False multiple transduction are generated (used for audio-based normalization)
+    """
+
+    def __init__(self, deterministic: bool = True):
+        super().__init__(name="time", kind="verbalize", deterministic=deterministic)
+
+        time_zone = convert_space(pynini.string_file(get_abs_path("data/time/time_zones.tsv")))
+        quoted_text = pynini.closure(NEMO_NOT_QUOTE)
+
+        def extract_field(field_name):
```

Comment:
move this to graph utils. it's enough of a general function that it should be cross module available. (probably should be adapted as a general method across tns but that's another beast)